### PR TITLE
New addon: Search bar on My Stuff

### LINKED ISSUE #1573: Search Your Projects/Studios
I currently have over 1,000 studios, and there's one studio I want to find, but since I can't really find a studio, I would like to be able to search my studios. This also might be useful for projects too.

### PR DESCRIPTION
**Pull request by lisa-wolfgang**
This is a highly requested feature.
Resolves #1573 (and its duplicate #2063)
Resolves [this request](https://discord.com/channels/806602307750985799/806602307750985803/841749392179986484), [this duplicate suggestion](https://discord.com/channels/806602307750985799/806609030935740477/842128989924622406), [this duplicate suggestion](https://discord.com/channels/806602307750985799/806609030935740477/854651769329156117), and [this duplicate suggestion](https://discord.com/channels/806602307750985799/806609030935740477/856275678881513473)
More or less resolves [this suggestion](https://discord.com/channels/806602307750985799/806609030935740477/894943914396635206), #1184, and the My Stuff portion of #3071

### Changes

image

When this addon is enabled, it adds a search bar on the My Stuff page next to the `Sort by` dropdown. The search bar is autofocused on page load, so the user can start typing their search as soon as the page loads. Beginning to type after unfocusing the search bar will also autofocus it.

When the user starts typing, the addon searches the items on the page and sorts them based on relevance to the given query. Results that are too irrelevant are excluded from the results. The results continue to update as the user continues typing their query. In addition, once the results have been calculated, pressing the `Enter` key will open the top search result.

When a search is performed and no results are returned, the addon will automatically click the "Load more" button and continue doing this until a new match is found or there are no more pages to load.

image

The addon will also do this after the user clicks the "Load more" button at the bottom of the search results.

image

The search bar is completely functional on all five My Stuff tabs: All Projects, Shared Projects, Unshared Projects, My Studios, and Trash. In addition, when the search bar is not used, the My Stuff page behaves the same as in vanilla Scratch, with one exception -- the `Hosted/curated by me` filter on the My Studios page has been removed because the Scratch API endpoint no longer supports it properly.

### Reason for changes

Users with many projects ~~such as W_L~~ can attest to the fact that finding projects in their My Stuff is quite time-consuming.
This feature greatly decreases the time it takes to locate a project, studio, or trash item on the page.

### Tests

Tested in browser w/ Chromium 94.
Compatible with and without the Scratch 2.0 -> 3.0 addon.

Known issues:
- Many of Scratch's API endpoints no longer function correctly, which means that many `Sort by` criteria, such as `Remixes` and `Comments`, don't return projects correctly. Before reporting issues with changing `Sort by` excluding certain results, first manually check that the issue also occurs in vanilla My Stuff.

### CODE REVIEW

**Comment by lisa-wolfgang** [comment]
> 
> 
> Instead of using fuse, it should sort by the value in scratch's "sort by" dropdown.

That wouldn't make sense, though. The point of a search bar is to order by relevance. If we didn't use something like Fuse, what would the point of this addon be?

**Comment by iqnite** [comment]
When will this come?

File: addons/scratchr2/mystuff.css [hunk 1]
```diff
@@ -156,3 +157,18 @@
...
+#search-my-stuff-input {
+  padding: 0 16px;
+  height: 48px;
+  font-size: 13px;
+}
+input:focus,
+textarea:focus {
+  box-shadow: 0 0 0 4px hsla(163, 85%, 40%, 0.25);
+  border-color: hsl(163, 85%, 40%);
+}
```
[the remark this answers is not in the document]
Comment (lisa-wolfgang) [reply addons/scratchr2/mystuff.css:174]: I think it came from the forum search bar addon. I will modify the selector to be more specific.

File: addons/search-my-stuff/userscript.js [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    // Determine which tab we're on and switch the placeholder text
+    if (window.location.href.includes("galleries")) {
+      searchBar.setAttribute("placeholder", msg("studio-placeholder"));
+    } else if (window.location.href.includes("trash")) {
+      searchBar.setAttribute("placeholder", msg("trash-placeholder"));
+    } else {
+      searchBar.setAttribute("placeholder", msg("project-placeholder"));
+    }
+    // Add new search elements
+    await searchDropdown;
```
[the remark this answers is not in the document]
Comment (lisa-wolfgang) [reply addons/search-my-stuff/userscript.js:96]: Should be an `addon.tab.waitForElement` -- I'll fix that.

**Comment by lisa-wolfgang** [comment]
> 
> 
> This behavior is kinda weird (I did not scroll up):

@WorldLanguages It's supposed to send you back to the top when new results are found in the event that better results would appear.

File: addons/search-my-stuff/addon.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "Search bar on My Stuff",
+  "description": "Allows you to search your projects, studios, and trash on the My Stuff page. Simply start typing anywhere on the page to start a search. You can also use Ctrl + Enter after searching to quickly open the top search result.",
+  "info": {
+    "type": "notice",
+    "text": "If you have a lot of projects or studios, the search might take a while. This is to avoid spamming the website with network requests. Please be patient.",
+    "id": "pleaseBePatient"
+  },
```
Comment (WorldLanguages) [inline addons/search-my-stuff/addon.json:8]: Whoops, this part of the manifest is invalid

**Pull request merged**